[TT-16831] docs: add OpenTelemetry metrics pages and restructure metrics section (#1808)
TykTechnologies/tyk-docs: api-management/observability/guides/best-practices.mdx

---
title: "Tyk Gateway Observability Best Practices"
description: "Production-ready guidance for configuring Tyk Gateway observability: OTLP export topology, metrics cardinality control, trace sampling strategy, and log collection."
keywords: "OpenTelemetry, OTel, Observability, Best Practices, Cardinality, Sampling, Metrics, Traces, Logs, OTLP, Prometheus, Grafana, Jaeger, Performance"
sidebarTitle: "Observability Best Practices"
---

> Tyk Gateway exports **metrics and traces** via the OpenTelemetry Protocol (OTLP) and writes **logs** to stdout/stderr for external collection.

Reviewer: ```suggestion
Tyk Gateway exports **metrics and traces** via the OpenTelemetry Protocol (OTLP) and writes **logs** to stderr for external collection.
```

This page covers production-ready configuration for each signal: which export topology to use, how to control metrics cardinality, how to tune trace sampling, and how to correlate logs with traces.

## Exporting Data via OTLP

> Use OTLP as the export protocol. It is vendor-neutral and supported by every modern observability backend. The gateway supports both gRPC (default, more efficient for high throughput) and HTTP transports. Choose gRPC unless your network or backend requires HTTP.

Reviewer: ```suggestion
Use OTLP as the export protocol. It is vendor-neutral and supported by every modern observability backend. The Gateway supports both gRPC (default, more efficient for high throughput) and HTTP transports. Choose gRPC unless your network or backend requires HTTP.
```

### Open Source Backends

| Backend | Best for | Notes |
|---------|----------|-------|
| **Grafana LGTM** (Loki, Grafana, Tempo, Mimir) | All signals | Tempo for traces, Mimir/Prometheus for metrics, Loki for logs |
| **Jaeger** | Tracing only | Accepts OTLP natively since v1.35; simple to self-host |
| **Prometheus + Grafana** | Metrics only | Pull model; use OTel Collector's Prometheus exporter or `remote_write` |
| **ELK / OpenSearch** | Logs + APM | Elastic APM accepts OTLP; Logstash can ingest OTel Collector output |

For vendor-specific configuration (Datadog, Dynatrace, New Relic, Elastic, Jaeger), see the [Traces](/api-management/traces) configuration guide.

## Metrics: Cardinality and Performance

### What Is the Cardinality Problem?

Each unique combination of dimension values for a metric creates a separate time series. Unbounded dimensions, such as one label per user, per IP address, or per request ID, cause exponential growth in series count. This consumes memory in Tyk Gateway, increases storage costs in your backend, and slows query performance.

### Tyk's Built-In Cardinality Limit

> Tyk caps each metric instrument at **2,000 unique label-value combinations** by default, controlled by the `cardinality_limit` field in the [OpenTelemetry metrics configuration](/api-management/metrics/opentelemetry-configuration#cardinality-control). When the limit is exceeded, additional combinations are aggregated into an overflow bucket marked with the attribute `otel.metric.overflow=true`. The aggregate counts are preserved in the overflow bucket, but you lose the ability to break down data by the overflowing dimension combination.

Reviewer: Broken link - I think it should be to: 

```suggestion
Tyk ordinarily caps each metric instrument at **2,000 unique label-value combinations**. A different limit can be set using the [cardinality control](/api-management/logs-metrics#cardinality-control).

If the configured limit is exceeded, the additional combinations are aggregated into an overflow bucket marked with the attribute `otel.metric.overflow=true`. Aggregate counts are preserved in the overflow bucket, but you lose the ability to break down data by the overflowing dimension combination.
```

> Alert on overflow to catch cardinality issues early:

Reviewer: Is this only for Prometheus?

Reviewer 2: the otel.metric.overflow=true attribute is set by the OTel SDK itself. The metric is available for all the time-series storage options. The quering options may vary, the below example is for promql

```promql
increase(<your_metric_name>{otel_metric_overflow="true"}[5m]) > 0
```

> The limit can be raised, but the correct fix is reducing cardinality in your [custom metrics configuration](/api-management/metrics/custom-metrics).

Reviewer: ```suggestion
If you are hitting the cardinality cap then the limit can be raised, but the recommended action is to reduce cardinality in your [custom metrics configuration](/api-management/metrics/custom-metrics).
```

The OTel SDK processes metrics on a fast path when an instrument has 10 or fewer dimensions. Exceeding this threshold increases memory allocations and slows metric recording. Keep each custom metric instrument to **10 or fewer dimensions**.

### Dimension Safety Guide

When defining [custom metrics](/api-management/metrics/custom-metrics), choose dimension sources based on their cardinality characteristics:

| Cardinality | Sources | Examples |
|-------------|---------|---------|
| **Safe** | `metadata`, `config_data` bounded fields | `listen_path`, `endpoint`, `method`, `http.response.status_code` |
| **Caution** | `session` fields, bounded JWT claims | `api_key`, `oauth_id`, `alias` (bounded per tenant but can be large) |
| **Avoid** | `header` / `context` with unbounded values | `ip_address`, `user_id` JWT claim, `request_id`, raw `path` |
| **Never** | Token or bearer values | Unique per request; exhausts the cardinality limit immediately |

## Traces: Sampling Strategy

Trace data is the most expensive observability signal to collect, store, and query. Sampling controls what fraction of traces you keep.

### Head-Based Sampling

> Configured via `opentelemetry.traces.sampling.type` and `opentelemetry.traces.sampling.rate` in the [gateway configuration](/api-management/traces#sampling-strategies). The sampling decision is made at the start of each trace, before any span data exists.

Reviewer: ```suggestion
With *head-based* sampling, the decision whether to create a trace for a request is made before any span data exists. In other words, the sampling logic is applied in the Gateway. This is controlled using the `TraceIDRatioBased` approach as described in the [trace sampling](/api-management/traces#sampling) section.
```

```json
{
  "opentelemetry": {
    "traces": {
      "sampling": {
        "type": "TraceIDRatioBased",
        "rate": 0.1
      }
    }
  }
}
```

**Characteristics:**
- Fast: no buffering required, predictable overhead
- Limitation: cannot guarantee capture of all errors or slow outliers at low sample rates
- **Recommended default: 10% (`rate: 0.1`)** for most production deployments

> ### Tail-Based Sampling via OTel Collector

Reviewer: ```suggestion
### Tail-Based Sampling
```

> The sampling decision is made after the full trace is collected. The OTel Collector buffers spans in memory, then applies policies to decide what to keep. This allows you to guarantee 100% retention of error traces and slow requests regardless of the overall sample rate.

Reviewer: ```suggestion
With *tail-based* sampling, the sampling decision is made in the OTel Collector after the full trace has been collected. The OTel Collector buffers spans in memory, then applies policies to decide what to keep. This allows you to guarantee 100% retention of error traces and slow requests regardless of the overall sample rate.
```

> To use tail-based sampling, send all traces from Tyk to the Collector (set `type: AlwaysOn`, which is the default), then apply policies in the Collector:

Reviewer: ```suggestion
To use tail-based sampling, send all traces from Tyk to the Collector using the `AlwaysOn` approach as described in the [trace sampling](/api-management/traces#sampling) section, then apply policies in the Collector. This is the default setting.
```

```yaml
# OTel Collector processors section
processors:
  tail_sampling:
    decision_wait: 10s
    policies:
      - name: keep-errors
        type: status_code
        status_code: {status_codes: [ERROR]}
      - name: keep-slow
        type: latency
        latency: {threshold_ms: 1000}
      - name: sample-rest
        type: probabilistic
        probabilistic: {sampling_percentage: 5}
```

### Sampling Trade-Offs

| Strategy | Cost | Accuracy | Best for |
|----------|------|----------|---------|
| `AlwaysOn` (100%) | High | Perfect | Dev/staging, low-traffic APIs |
| `TraceIDRatioBased`, `rate: 0.1` (10%) | Low | Good average | Most production deployments |
| `TraceIDRatioBased`, `rate: 0.01` (1%) | Very low | Poor for rare events | Very high traffic, budget-constrained |
| Tail-based via OTel Collector | Medium (Collector infra) | Best | When you need all errors and outliers |

## Logs: Collection and Correlation

> Tyk Gateway writes logs to stdout/stderr, not via OTLP. They must be collected by an external agent.

Reviewer: ```suggestion
Tyk Gateway writes logs to stderr, not via OTLP. They must be collected by an external agent.
```

### Enable Access Logs

> Access logs record one line per request: HTTP method, path, upstream latency, status code, response size, and client IP. They are the fastest way to observe gateway traffic without a full tracing setup and are essential for log-based alerting such as 5xx spike detection.

Reviewer: ```suggestion
Access logs record one line per request: HTTP method, path, upstream latency, status code, response size, and client IP. They are the fastest way to observe Gateway traffic without a full tracing setup and are essential for log-based alerting such as 5xx spike detection.
```

> Enable in the gateway configuration:

Reviewer: ```suggestion
Access logs are enabled in the Gateway configuration or the equivalent [environment variable](/tyk-oss-gateway/configuration#access_logs-enabled):
```

```json
{
  "access_logs": {
    "enabled": true
  }
}
```

See [Access Logs](/api-management/logs#access-logs) for the full list of configurable fields and custom template options.

### Use JSON Log Format

> JSON format has lower parsing overhead than the default text format and is directly indexable by log backends (Loki, Elasticsearch, CloudWatch). Enable it with:

Reviewer: ```suggestion
JSON format has lower parsing overhead than the default text format and is directly indexable by log backends (Loki, Elasticsearch, CloudWatch). Enable this mode in the Gateway configuration or the equivalent [environment variable](/tyk-oss-gateway/configuration#log_format):
```

```json
{
  "log_format": "json"
}
```

> Recommended for all production deployments.

Reviewer: ```suggestion
JSON log format is recommended for all production deployments.
```
